packages/macros/src/ember-addon-main.ts

```diff
@@ -70,6 +72,30 @@ export = {
       let appInstance = this._findHost();
       let source = appOrAddonInstance.root || appOrAddonInstance.project.root;
       babelPlugins.unshift(MacrosConfig.for(appInstance).babelPluginConfig(source));
+
+      let yarnLockPath = join(appInstance.project.root, 'yarn.lock');
+      let npmLockPath = join(appInstance.project.root, 'package-lock.json');
+      let packagePath = join(appInstance.project.root, 'package.json');
+      let lockFileBuffer;
+
+      if (fs.existsSync(yarnLockPath)) {
+        lockFileBuffer = fs.readFileSync(yarnLockPath);
+      } else if (fs.existsSync(npmLockPath)) {
+        lockFileBuffer = fs.readFileSync(npmLockPath);
+      } else {
+        // no lock file found, using package.json as a fall back
+        lockFileBuffer = fs.readFileSync(packagePath);
```

Comment:
If we always include package.json, we can encourage our users to use `link:` (yarn) or `file:` (npm) to perform linking, as that would at-least make that level cache aware.

If we extended that further, and always invalided the cache for `link:` or `file:` if they are on npm, we could implicitly prevent deeply nested linked deps that themselves later get more links from causing cache corruption. TL;DR if this is the path we deem appropriate it would result in a single linked dependency disabling babel caching. This is pretty conservative... but may be appropriate. We should think through this

Reply:
Ya, this seems good to me.

Reply:
FWIW, this also seems reasonable for the original purposes in #841. Can we move this logic over into `@embroider/utils` and use it from `@embroider/core` for #841's use case and also use it here for this `dependencySatisfies` invalidation?

Reply:
Both packages (macros and core) use `@embroider/shared-internals` which is probably the best home for this share plugin.

Reply:
Yes, `@embroider/utils` is utilities for app & addons to help with embroider compatibility. `@embroider/shared-internals` is for this sort of thing.